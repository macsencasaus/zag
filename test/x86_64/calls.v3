// RUN: %vector -stdout %s > /tmp/tmp.bin && objdump -d -M intel /tmp/tmp.bin

fn id(x: i32) i32 {
    return x;
}

fn call_id(x: i32) i32 {
    return id(x);
}

fn factorial(n: u32) u32 {
    if (n == 0)
        return 1;
    return n * factorial(n - 1);
}

fn factorial_tail(n: u32, res: u32) u32 {
    if (n == 0)
        return res;
    return factorial_tail(n - 1, n * res);
}

fn func_ptrs(func: fn(i32) i32) i32 {
    return func(2);
}

fn direct_to_indirect() i32 {
    var id_fn = id;
    return id_fn(1);
}
