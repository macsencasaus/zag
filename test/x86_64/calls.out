
/tmp/tmp.bin:     file format elf64-x86-64


Disassembly of section .text:

0000000000000000 <id>:
   0:	55                   	push   rbp
   1:	48 89 e5             	mov    rbp,rsp
   4:	48 83 ec 10          	sub    rsp,0x10
   8:	89 7d fc             	mov    DWORD PTR [rbp-0x4],edi
   b:	8b 45 fc             	mov    eax,DWORD PTR [rbp-0x4]
   e:	c9                   	leave
   f:	c3                   	ret

0000000000000010 <call_id>:
  10:	55                   	push   rbp
  11:	48 89 e5             	mov    rbp,rsp
  14:	48 83 ec 10          	sub    rsp,0x10
  18:	89 7d fc             	mov    DWORD PTR [rbp-0x4],edi
  1b:	8b 7d fc             	mov    edi,DWORD PTR [rbp-0x4]
  1e:	e8 00 00 00 00       	call   23 <call_id+0x13>
  23:	89 45 f8             	mov    DWORD PTR [rbp-0x8],eax
  26:	8b 45 f8             	mov    eax,DWORD PTR [rbp-0x8]
  29:	c9                   	leave
  2a:	c3                   	ret

000000000000002b <factorial>:
  2b:	55                   	push   rbp
  2c:	48 89 e5             	mov    rbp,rsp
  2f:	48 83 ec 10          	sub    rsp,0x10
  33:	89 7d fc             	mov    DWORD PTR [rbp-0x4],edi
  36:	8b 45 fc             	mov    eax,DWORD PTR [rbp-0x4]
  39:	b9 00 00 00 00       	mov    ecx,0x0
  3e:	39 c8                	cmp    eax,ecx
  40:	0f 94 c0             	sete   al
  43:	0f b6 c0             	movzx  eax,al
  46:	89 45 f8             	mov    DWORD PTR [rbp-0x8],eax
  49:	8b 45 f8             	mov    eax,DWORD PTR [rbp-0x8]
  4c:	85 c0                	test   eax,eax
  4e:	0f 84 07 00 00 00    	je     5b <factorial+0x30>
  54:	b8 01 00 00 00       	mov    eax,0x1
  59:	c9                   	leave
  5a:	c3                   	ret
  5b:	8b 45 fc             	mov    eax,DWORD PTR [rbp-0x4]
  5e:	b9 01 00 00 00       	mov    ecx,0x1
  63:	29 c8                	sub    eax,ecx
  65:	89 45 f4             	mov    DWORD PTR [rbp-0xc],eax
  68:	8b 7d f4             	mov    edi,DWORD PTR [rbp-0xc]
  6b:	e8 00 00 00 00       	call   70 <factorial+0x45>
  70:	89 45 f8             	mov    DWORD PTR [rbp-0x8],eax
  73:	8b 45 fc             	mov    eax,DWORD PTR [rbp-0x4]
  76:	8b 4d f8             	mov    ecx,DWORD PTR [rbp-0x8]
  79:	f7 e9                	imul   ecx
  7b:	89 45 f0             	mov    DWORD PTR [rbp-0x10],eax
  7e:	8b 45 f0             	mov    eax,DWORD PTR [rbp-0x10]
  81:	c9                   	leave
  82:	c3                   	ret

0000000000000083 <factorial_tail>:
  83:	55                   	push   rbp
  84:	48 89 e5             	mov    rbp,rsp
  87:	48 83 ec 20          	sub    rsp,0x20
  8b:	89 7d fc             	mov    DWORD PTR [rbp-0x4],edi
  8e:	89 75 f8             	mov    DWORD PTR [rbp-0x8],esi
  91:	8b 45 fc             	mov    eax,DWORD PTR [rbp-0x4]
  94:	b9 00 00 00 00       	mov    ecx,0x0
  99:	39 c8                	cmp    eax,ecx
  9b:	0f 94 c0             	sete   al
  9e:	0f b6 c0             	movzx  eax,al
  a1:	89 45 f4             	mov    DWORD PTR [rbp-0xc],eax
  a4:	8b 45 f4             	mov    eax,DWORD PTR [rbp-0xc]
  a7:	85 c0                	test   eax,eax
  a9:	0f 84 05 00 00 00    	je     b4 <factorial_tail+0x31>
  af:	8b 45 f8             	mov    eax,DWORD PTR [rbp-0x8]
  b2:	c9                   	leave
  b3:	c3                   	ret
  b4:	8b 45 fc             	mov    eax,DWORD PTR [rbp-0x4]
  b7:	b9 01 00 00 00       	mov    ecx,0x1
  bc:	29 c8                	sub    eax,ecx
  be:	89 45 f0             	mov    DWORD PTR [rbp-0x10],eax
  c1:	8b 45 fc             	mov    eax,DWORD PTR [rbp-0x4]
  c4:	8b 4d f8             	mov    ecx,DWORD PTR [rbp-0x8]
  c7:	f7 e9                	imul   ecx
  c9:	89 45 ec             	mov    DWORD PTR [rbp-0x14],eax
  cc:	8b 7d f0             	mov    edi,DWORD PTR [rbp-0x10]
  cf:	8b 75 ec             	mov    esi,DWORD PTR [rbp-0x14]
  d2:	e8 00 00 00 00       	call   d7 <factorial_tail+0x54>
  d7:	89 45 f4             	mov    DWORD PTR [rbp-0xc],eax
  da:	8b 45 f4             	mov    eax,DWORD PTR [rbp-0xc]
  dd:	c9                   	leave
  de:	c3                   	ret

00000000000000df <func_ptrs>:
  df:	55                   	push   rbp
  e0:	48 89 e5             	mov    rbp,rsp
  e3:	48 83 ec 10          	sub    rsp,0x10
  e7:	48 89 7d f8          	mov    QWORD PTR [rbp-0x8],rdi
  eb:	bf 02 00 00 00       	mov    edi,0x2
  f0:	48 8b 45 f8          	mov    rax,QWORD PTR [rbp-0x8]
  f4:	ff 10                	call   QWORD PTR [rax]
  f6:	89 45 f4             	mov    DWORD PTR [rbp-0xc],eax
  f9:	8b 45 f4             	mov    eax,DWORD PTR [rbp-0xc]
  fc:	c9                   	leave
  fd:	c3                   	ret

00000000000000fe <direct_to_indirect>:
  fe:	55                   	push   rbp
  ff:	48 89 e5             	mov    rbp,rsp
 102:	48 83 ec 10          	sub    rsp,0x10
 106:	48 8d 05 00 00 00 00 	lea    rax,[rip+0x0]        # 10d <direct_to_indirect+0xf>
 10d:	48 89 45 f8          	mov    QWORD PTR [rbp-0x8],rax
 111:	bf 01 00 00 00       	mov    edi,0x1
 116:	48 8b 45 f8          	mov    rax,QWORD PTR [rbp-0x8]
 11a:	ff 10                	call   QWORD PTR [rax]
 11c:	89 45 f4             	mov    DWORD PTR [rbp-0xc],eax
 11f:	8b 45 f4             	mov    eax,DWORD PTR [rbp-0xc]
 122:	c9                   	leave
 123:	c3                   	ret
